parsl/tests/test_providers/test_local_provider.py

```diff
@@ -0,0 +1,202 @@
+import os
+import random
+import shutil
+import socket
+import subprocess
+import tempfile
+import threading
+import time
+
+from parsl.channels import LocalChannel, SSHChannel
+from parsl.launchers import SingleNodeLauncher
+from parsl.providers import LocalProvider
+from parsl.providers.provider_base import JobState
+
+
+def _run_tests(p: LocalProvider):
+    status = _run(p, '/bin/true')
+    assert status.message is None
+    assert status.state == JobState.COMPLETED
+    assert status.exit_code == 0
+    assert status.stdout == ''
+    assert status.stderr == ''
+
+    status = _run(p, '/bin/true', np=2)
+    assert status.message is None
+    assert status.state == JobState.COMPLETED
+    assert status.exit_code == 0
+    assert status.stdout == ''
+    assert status.stderr == ''
+
+    status = _run(p, '/bin/false')
+    assert status.state == JobState.FAILED
+    assert status.exit_code != 0
+    assert status.stdout == ''
+    assert status.stderr == ''
+
+    status = _run(p, '/bin/false', np=2)
+    assert status.state == JobState.FAILED
+    assert status.exit_code != 0
+    assert status.stdout == ''
+    assert status.stderr == ''
+
+    status = _run(p, '/bin/echo -n magic')
+    assert status.state == JobState.COMPLETED
+    assert status.exit_code == 0
+    assert status.stdout == 'magic'
+    assert status.stderr == ''
+
+    status = _run(p, '/bin/echo -n magic 1>&2')
+    assert status.state == JobState.COMPLETED
+    assert status.exit_code == 0
+    assert status.stdout == ''
+    assert status.stderr == 'magic'
+
+
+def test_local_channel():
+    with tempfile.TemporaryDirectory() as script_dir:
+        script_dir = tempfile.mkdtemp()
+        p = LocalProvider(channel=LocalChannel(), launcher=SingleNodeLauncher(debug=False))
+        p.script_dir = script_dir
+        _run_tests(p)
+
+
+SSHD_CONFIG = """
+Port {port}
+ListenAddress 127.0.0.1
+HostKey {hostkey}
+AuthorizedKeysFile {connpubkey}
+AuthenticationMethods publickey
+StrictModes no
+Subsystem sftp {sftp_path}
+"""
+
+
+# It would probably be better, when more formalized site testing comes into existence, to
+# use a site-testing provided server/configuration instead of the current scheme
+def test_ssh_channel():
+    with tempfile.TemporaryDirectory() as config_dir:
+        sshd_thread, priv_key, server_port = _start_sshd(config_dir)
+        try:
+            with tempfile.TemporaryDirectory() as remote_script_dir:
+                # run this once so we add host keys and whatnot
```

Comment:
`SSHChannel` has this set:

> self.ssh_client.set_missing_host_key_policy(paramiko.AutoAddPolicy())

That should be adding host keys. Does something else break here?

Reply:
Didn't try. I will shortly.

Reply:
Fails if the key file does not exist, but works if an empty one is created. I'll do that instead.